doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,339 @@ Applying theorems
+
+      In addition,
+      `apply` in a hypothesis isn't as flexible as
+      `apply` in the goal: for hypotheses, the unbound variable can be bound
+      by name (as shown) or values for all the variables can be given
+      positionally,
+      i.e. `apply Nat.le_trans with (x * x ^ n) (y * y ^ n) (y * x ^ n) in H.`
+
+      .. coqtop:: reset none
+
+         Require Import PeanoNat.
+         Require Import Nat.
+         Infix "^" := pow : nat_scope.
+         Goal forall (x y n : nat) (H: x * x ^ n <= y * y ^ n), True.
+
+      .. coqtop:: out
+
+         intros x y n H.
+
+      .. coqtop:: all
+
+         apply Nat.le_trans with (p := y * x ^ n) in H.
+
+   .. _apply_with_iff:
+   .. example:: Applying theorems with `<->`
+
+      Theorems that use `<->` to state a logical equivalence behave
+      consistently when applied to goals and hypotheses; the tactic
+      picks the useful alternative.  The lemma
+      is `iff: 1 = 1 <-> 2 = 2`:
+
+      .. coqtop:: reset none
+
+         Lemma iff: 1 = 1 <-> 2 = 2.  Admitted.
+         Goal forall (H: 1 = 1), 1 = 1.
+
+      .. coqtop:: out
+
+         intros H.
+
+      .. coqtop:: all
+
+         apply iff.
+         apply iff in H.
+
+   .. tacn:: eapply {+, {? > } @one_term {? with @bindings } } {? @in_hyp_as }
+
+      Behaves like :tacn:`apply`, but creates
+      :ref:`existential variables <Existential-Variables>`
+      when Coq is unable to deduce instantiations for variables, rather than failing.
+
+   .. tacn:: rapply @one_term
+
+      Behaves like :tacn:`eapply` but
+      uses the proof engine of :tacn:`refine` to handle
+      existential variables, holes and conversion problems.  This may
       result in slightly different behavior regarding which conversion
       problems are solvable.  However, like :tacn:`apply` but unlike
-      :tacn:`eapply`, :tacn:`rapply` will fail if there are any holes
-      which remain in :n:`@term` itself after typechecking and
+      :tacn:`eapply`, :tacn:`rapply` fails if any holes remain
+      in :n:`@one_term` itself after typechecking and
       typeclass resolution but before unification with the goal.  More
-      technically, :n:`@term` is first parsed as a
+      technically, :n:`@one_term` is first interpeted as a
       :production:`constr` rather than as a :production:`uconstr` or
       :production:`open_constr` before being applied to the goal. Note
-      that :tacn:`rapply` prefers to instantiate as many hypotheses of
-      :n:`@term` as possible.  As a result, if it is possible to apply
-      :n:`@term` to arbitrarily many arguments without getting a type
+      that :tacn:`rapply` tries to instantiate as many hypotheses of
+      :n:`@one_term` as possible.  As a result, if it is possible to apply
+      :n:`@one_term` to arbitrarily many arguments without getting a type
       error, :tacn:`rapply` will loop.
 
+      .. todo PR: reword sentence above to remove :production:, don't understand
+         that sentence
```

Comment:
Basically, it says that the `@one_term` is fully typechecked and cannot contain leftover holes before the tactic tries to match it with the goal, whereas `eapply` would be OK with that (it accepts "open terms", see https://coq.inria.fr/refman/user-extensions/syntax-extensions.html#tactic-notations and https://coq.inria.fr/refman/proof-engine/ltac2.html#built-in-quotations for place where the notions of `uconstr` and `open_constr` are mentioned).

However, it is false when it says "like apply but unlike apply". `rapply` is actually unlike both `appply` and `eapply` for this aspect (both accept holes in the `one_term` argument if they can match them infer them from the conclusion of the goal).

Reply:
Maybe I should just drop the "More technically ..." sentence here.  The sentence before it seems clear.  If kept, it deserves added hyperlinks.

`open_constr` is not listed in `Tactic Notation`, I can add it.  Also, perhaps that section should better explain the difference between it `constr` and `uconstr` (which is ??).

What's the definition of "open term"?  Does it mean that wildcards are allowed, or something else?

Reply:
If I'm not mistaken (to check, possibly by asking for other reviewers in a separate PR):
- `constr` means fully elaborated term (it has no holes and it has a type)
- `open_constr` means not-fully elaborated term (it can have holes represented by wildcards or, more generally, existential variables)
- `uconstr` means untyped term, so the tactic or tactic notation would accept its argument without running type checking at all, and could type check it in a latter phase in an extended context, or not check it before `Qed`.

Reply:
WDYT about my question above?

> Maybe I should just drop the "More technically ..." sentence here. The sentence before it seems clear. If kept, it deserves added hyperlinks.

Reply:
Yes, I'd personally be fine with dropping this.